File: framework/languages/node/e2e/PubSub/Client/Scenarios/ps-a1-fanout-basic-delivery-scenario.ts
// PS-A1: 준비된 subscriber가 같은 event를 받는다 시나리오를 검증한다.
import { randomUUID } from 'node:crypto';
import { PubSubNames, type EvidenceWaitReq } from '../../Shared/messages';
import { commonContiguousSequence } from '../Support/evidence';
import { postJson, postJsonWithin, postStatus } from '../../../http-client';
import { delay, ensure } from '../Support/scenario-assert';
import { waitForEvent } from '../Support/subscriber-observation';

export async function runPsA1(publisher: string, subscribers: readonly string[]): Promise<void> {
  const runId = randomUUID().replaceAll('-', '');
  for (let i = 1; i <= 120; i += 1) {
    await publishEvent(publisher, PubSubNames.mainTopic, runId, i, `warmup-${i}`);
    if (i % 10 === 0) {
      await delay(25);
    }
  }
  await waitForAll(subscribers, { containsAll: ['event|', `run=${runId}`, `topic=${PubSubNames.mainTopic}`], timeoutMilliseconds: 10_000 });

  const measureStart = 1000;
  const measureCount = 12;
  for (let i = measureStart; i < measureStart + measureCount; i += 1) {
    await publishEvent(publisher, PubSubNames.mainTopic, runId, i, `measure-${i}`);
  }
  const snapshots = await waitForAll(subscribers, {
    containsAll: ['event|', `run=${runId}`, `topic=${PubSubNames.mainTopic}`],
    containsAllLineGroups: [0, 1, 2].map((offset) => [
      `seq=${measureStart + offset}|`,
      `value=measure-${measureStart + offset}`
    ]),
    timeoutMilliseconds: 10_000
  });
  ensure(
    commonContiguousSequence(
      snapshots,
      runId,
      PubSubNames.mainTopic,
      measureStart,
      measureStart + measureCount - 1,
      'measure-'
    ).length >= 3,
    'PS-A1 expected common contiguous sequence on all subscribers.'
  );
  console.log('scenario PS-A1 passed');
}

export async function publishEvent(publisher: string, topic: string, runId: string, sequence: number, value: string): Promise<void> {
  await postJson(publisher, '/publish/event', { topic, runId, sequence, value });
}

export async function publishUntilDelivered(
  publisher: string,
  subscriber: string,
  topic: string,
  sequence: number,
  value: string,
  timeoutMilliseconds = 20_000
): Promise<string> {
  const deadline = Date.now() + timeoutMilliseconds;
  while (Date.now() < deadline) {
    const runId = randomUUID().replaceAll('-', '');
    await publishEvent(publisher, topic, runId, sequence, value);
    try {
      await waitForEvent(subscriber, runId, sequence, value, 750);
      return runId;
    } catch {
      await delay(100);
    }
  }
  throw new Error(`Timed out waiting for fanout delivery of value '${value}'.`);
}

export async function publishUntilAllDelivered(
  publisher: string,
  subscribers: readonly string[],
  topic: string,
  sequence: number,
  value: string,
  timeoutMilliseconds = 20_000
): Promise<string> {
  const deadline = Date.now() + timeoutMilliseconds;
  while (Date.now() < deadline) {
    const runId = randomUUID().replaceAll('-', '');
    await Promise.all(subscribers.map((subscriber) => postStatus(`${subscriber}/evidence/clear`)));
    await publishEvent(publisher, topic, runId, sequence, value);
    try {
      await waitForAll(subscribers, {
        containsAll: ['event|', `run=${runId}`, `topic=${topic}`],
        timeoutMilliseconds: 750
      });
      return runId;
    } catch {
      await delay(100);
    }
  }
  throw new Error(`Timed out waiting for all fanout subscribers to receive value '${value}'.`);
}

export async function waitForAll(subscribers: readonly string[], request: EvidenceWaitReq): Promise<readonly (readonly string[])[]> {
  const timeoutMilliseconds = request.timeoutMilliseconds ?? 10_000;
  return await Promise.all(subscribers.map((subscriber) => postJsonWithin<readonly string[]>(
    subscriber,
    '/evidence/wait',
    request,
    timeoutMilliseconds + 1_000
  )));
}

File: framework/languages/dotnet/e2e/PubSub/Client/Scenarios/PsA1FanoutBasicDeliveryScenario.cs
// Verifies PS-A1 Fanout Basic Delivery behavior.
using PubSub.Client.Support;
using PubSub.Shared;
using Zlink.HttpClient;

namespace PubSub.Client.Scenarios;

// PS-A1: verifies that one publisher fanouts a shared event range to every subscriber.
internal static class PsA1FanoutBasicDeliveryScenario
{
    public static async Task RunAsync(ZLinkHttpClient publisher, IReadOnlyList<ZLinkHttpClient> subscribers)
    {
        var runId = Guid.NewGuid().ToString("N");
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", $"activate-{runId}")
            .Query("sequence", "0")
            .Query("value", "activate-publisher")
            .AsyncRaw();
        await Task.WhenAll(subscribers.Select(subscriber =>
            SubscriberObservation.WaitForEventAsync(subscriber, $"activate-{runId}", 0)));

        var measureStart = 100;
        var measureCount = 12;

        // The measured range is separate from warm-up so the fanout oracle uses only post-barrier data.
        for (var i = measureStart; i < measureStart + measureCount; i++)
            await publisher.Post("/publish/event")
                .Query("topic", PubSubNames.MainTopic)
                .Query("runId", runId)
                .Query("sequence", i.ToString())
                .Query("value", $"measure-{i}")
                .AsyncRaw();

        // Publish submit does not promise remote delivery, so the scenario checks a shared contiguous range.
        var measuredSnapshots = await WaitForAllSubscribersAsync(
            subscribers,
            new EvidenceWaitReq(
                ["event|", $"run={runId}", $"topic={PubSubNames.MainTopic}"],
                [
                    [$"seq={measureStart}|"],
                    [$"seq={measureStart + 1}|"],
                    [$"seq={measureStart + 2}|"]
                ]));
        ZlinkStreamAssert.Ensure(
            Evidence.CommonContiguousSequence(
                measuredSnapshots,
                runId,
                PubSubNames.MainTopic,
                measureStart,
                measureStart + measureCount - 1).Count >= 3,
            "PS-A1 expected common contiguous sequence on all subscribers.");

        Console.WriteLine("scenario PS-A1 passed");
    }

    private static async Task<string[][]> WaitForAllSubscribersAsync(
        IReadOnlyList<ZLinkHttpClient> subscribers,
        EvidenceWaitReq request)
    {
        var waits = subscribers
            .Select(subscriber => subscriber.Post("/evidence/wait").Body(request).Async<string[]>().AsTask())
            .ToArray();
        var responses = await Task.WhenAll(waits);
        return responses.Select(response => response.Body).ToArray();
    }
}
